### PR TITLE
Zunum branch merge

### PR DESCRIPTION
**Pull request by cmcmillan8**
# Hybrid files
 * Modified battery ducted fan energy network
 * Added serial hybrid ducted fan energy network
 * Added zero fidelity generator energy converter

### CODE REVIEW

File: regression/scripts/ducted_fan/battery_ducted_fan_network.py [hunk 1]
```diff
@@ -0,0 +1,269 @@
...
+#   Imports
+# ----------------------------------------------------------------------
+
+import SUAVE
+from SUAVE.Core import Units
+
+import numpy as np
+import pylab as plt
+
+import copy, time
```
Comment (planes) [inline regression/scripts/ducted_fan/battery_ducted_fan_network.py]: Do you need copy, time or container?

File: regression/scripts/ducted_fan/battery_ducted_fan_network.py [hunk 2]
```diff
@@ -0,0 +1,269 @@
...
+def energy_network():
+    # ------------------------------------------------------------------
+    #   Battery Ducted Fan Network
+    # ------------------------------------------------------------------    
+    
+    #instantiate the ducted fan network
+    ducted_fan = SUAVE.Components.Energy.Networks.Ducted_Fan()
+    ducted_fan.tag = 'ducted fan'
+
+    ducted_fan.number_of_engines = 2
```
Comment (planes) [inline regression/scripts/ducted_fan/battery_ducted_fan_network.py]: redundant with the next lines

File: regression/scripts/ducted_fan/battery_ducted_fan_network.py [hunk 3]
```diff
@@ -0,0 +1,269 @@
...
+ 
+    #total design thrust (includes all the engines)
+    thrust.total_design        = 2*700. * Units.lbf 
+    
+    
+    # add to network
+    ducted_fan.thrust          = thrust   
+  
+    numerics                   = Data()    
+    eta                        = 1.0
```
Comment (planes) [inline regression/scripts/ducted_fan/battery_ducted_fan_network.py]: what is eta?

File: regression/scripts/ducted_fan/battery_ducted_fan_network.py [hunk 4]
```diff
@@ -0,0 +1,269 @@
...
+    thrust.tag                 ='thrust'
+ 
+    #total design thrust (includes all the engines)
+    thrust.total_design        = 2*700. * Units.lbf 
+    
+    
+    # add to network
+    ducted_fan.thrust          = thrust   
+  
+    numerics                   = Data()
```
Comment (planes) [inline regression/scripts/ducted_fan/battery_ducted_fan_network.py]: is this numerics used?

File: regression/scripts/ducted_fan/battery_ducted_fan_network.py [hunk 5]
```diff
@@ -0,0 +1,269 @@
...
+    altitude      = 0.0 * Units.km
+    mach_number   = 0.01
+    isa_deviation = 0.
+    
+    #size the ducted fan
+    ducted_fan_sizing(ducted_fan,mach_number,altitude)
+    
+    
+    battery_ducted_fan                      = SUAVE.Components.Energy.Networks.Battery_Ducted_Fan()
+    battery_ducted_fan.tag                  = 'turbofan'
```
Comment (planes) [inline regression/scripts/ducted_fan/battery_ducted_fan_network.py]: no it's not

File: regression/scripts/ducted_fan/serial_hybrid_ducted_fan_network.py [hunk 6]
```diff
@@ -0,0 +1,276 @@
...
+#   Imports
+# ----------------------------------------------------------------------
+
+import SUAVE
+from SUAVE.Core import Units
+
+import numpy as np
+import pylab as plt
+
+import copy, time
```
Comment (planes) [inline regression/scripts/ducted_fan/serial_hybrid_ducted_fan_network.py]: imports

File: regression/scripts/ducted_fan/serial_hybrid_ducted_fan_network.py [hunk 7]
```diff
@@ -0,0 +1,276 @@
...
+def energy_network():
+    # ------------------------------------------------------------------
+    #   Battery Ducted Fan Network
+    # ------------------------------------------------------------------    
+    
+    #instantiate the ducted fan network
+    ducted_fan = SUAVE.Components.Energy.Networks.Ducted_Fan()
+    ducted_fan.tag = 'ducted fan'
+
+    ducted_fan.number_of_engines = 2
```
Comment (planes) [inline regression/scripts/ducted_fan/serial_hybrid_ducted_fan_network.py]: number of engines again

File: regression/scripts/ducted_fan/serial_hybrid_ducted_fan_network.py [hunk 8]
```diff
@@ -0,0 +1,276 @@
...
+    altitude      = 0.0 * Units.km
+    mach_number   = 0.01
+    isa_deviation = 0.
+    
+    #size the ducted fan
+    ducted_fan_sizing(ducted_fan,mach_number,altitude)
+    
+    
+    battery_ducted_fan                      = SUAVE.Components.Energy.Networks.Serial_Hybrid_Ducted_Fan()
+    battery_ducted_fan.tag                  = 'turbofan'
```
Comment (planes) [inline regression/scripts/ducted_fan/serial_hybrid_ducted_fan_network.py]: same

File: regression/scripts/ducted_fan/serial_hybrid_ducted_fan_network.py [hunk 9]
```diff
@@ -0,0 +1,276 @@
...
+    thrust.tag                 ='thrust'
+ 
+    #total design thrust (includes all the engines)
+    thrust.total_design        = 2*700. * Units.lbf 
+    
+    
+    # add to network
+    ducted_fan.thrust          = thrust   
+  
+    numerics                   = Data()
```
Comment (planes) [inline regression/scripts/ducted_fan/serial_hybrid_ducted_fan_network.py]: same

File: regression/scripts/ducted_fan/serial_hybrid_ducted_fan_network.py [hunk 10]
```diff
@@ -0,0 +1,276 @@
...
+    ducted_fan_sizing(ducted_fan,mach_number,altitude)
+    
+    
+    battery_ducted_fan                      = SUAVE.Components.Energy.Networks.Serial_Hybrid_Ducted_Fan()
+    battery_ducted_fan.tag                  = 'turbofan'
+    battery_ducted_fan.nacelle_diameter     = ducted_fan.nacelle_diameter
+    battery_ducted_fan.areas                = Data()
+    battery_ducted_fan.areas.wetted         = ducted_fan.areas.wetted
+    battery_ducted_fan.engine_length        = ducted_fan.engine_length
+    battery_ducted_fan.number_of_engines    = ducted_fan.number_of_engines
```
Comment (planes) [inline regression/scripts/ducted_fan/serial_hybrid_ducted_fan_network.py]: Wait is the number of engines multiplied twice? Once for the ducted fan and once for the battery ducted fan?

File: trunk/SUAVE/Components/Energy/Converters/Generator_Zero_Fid.py [hunk 11]
```diff
@@ -0,0 +1,88 @@
...
+
+# suave imports
+import SUAVE
+
+# package imports
+from SUAVE.Components.Energy.Energy_Component import Energy_Component
+import numpy as np
+
+# ----------------------------------------------------------------------
+#  Motor Class
```
Comment (planes) [inline trunk/SUAVE/Components/Energy/Converters/Generator_Zero_Fid.py]: Not a motor

File: trunk/SUAVE/Components/Energy/Converters/Generator_Zero_Fid.py [hunk 12]
```diff
@@ -0,0 +1,88 @@
...
+            sfc                [g/(KW*h)]
+            max_power          [W]
+
+        """       
+        
+        # Unpack
+        sfc         = self.sfc
+        max_power   = self.max_power
+
+        mdot = (max_power/1000) * (sfc/(3600*1000))
```
Comment (planes) [inline trunk/SUAVE/Components/Energy/Converters/Generator_Zero_Fid.py]: let's use some units for this

File: trunk/SUAVE/Components/Energy/Converters/Generator_Zero_Fid.py [hunk 13]
```diff
@@ -0,0 +1,88 @@
...
+        N/A
+    
+        Outputs:
+        self.outputs.
+            power_generated    [W]
+            mdot               [kg/s]
+    
+        Properties Used:
+        self.
+            sfc                [g/(KW*h)]
```
Comment (planes) [inline trunk/SUAVE/Components/Energy/Converters/Generator_Zero_Fid.py]: grams... we use kg's in SUAVE

File: trunk/SUAVE/Components/Energy/Networks/Battery_Ducted_Fan.py [hunk 14]
```diff
@@ -55,14 +55,20 @@ def __defaults__(self):
         self.propulsor        = None
         self.battery          = None
         self.motor_efficiency = .95
```
Comment (planes) [inline trunk/SUAVE/Components/Energy/Networks/Battery_Ducted_Fan.py]: default this to 0

File: trunk/SUAVE/Components/Energy/Networks/Battery_Ducted_Fan.py [hunk 15]
```diff
@@ -76,48 +82,65 @@ def evaluate_thrust(self,state):
...
+
+        # Calculate ducted fan power
+        results             = propulsor.evaluate_thrust(state)
+        propulsive_power    = results.power
+        motor_power         = propulsive_power/self.motor_efficiency 
+
+        # Run the ESC
+        esc.inputs.voltagein = self.voltage
+        esc.voltageout(conditions)
+        esc.inputs.currentout =  np.transpose(motor_power/np.transpose(esc.outputs.voltageout))
```
Comment (planes) [inline trunk/SUAVE/Components/Energy/Networks/Battery_Ducted_Fan.py]: maybe just transpose the motor power up above?

File: trunk/SUAVE/Components/Energy/Networks/Ducted_Fan.py [hunk 16]
```diff
@@ -3,6 +3,7 @@
 # 
 # Created:  Feb 2016, M. Vegh
 # Modified: Aug 2017, E. Botero
+#           Apr 2019, C. McMillan
```
Comment (planes) [inline trunk/SUAVE/Components/Energy/Networks/Ducted_Fan.py]: what did you change to justify putting your name on?

File: trunk/SUAVE/Components/Energy/Networks/Serial_Hybrid_Ducted_Fan.py [hunk 17]
```diff
@@ -0,0 +1,158 @@
...
+            Outputs:
+            None
+    
+            Properties Used:
+            N/A
+        """         
+        
+        self.propulsor        = None
+        self.battery          = None
+        self.motor_efficiency = .95
```
Comment (planes) [inline trunk/SUAVE/Components/Energy/Networks/Serial_Hybrid_Ducted_Fan.py]: default to 0

File: trunk/SUAVE/Components/Propulsors/Propulsor.py [hunk 18]
```diff
@@ -50,6 +50,16 @@ def __defaults__(self):
                 N/A
         """
         self.tag = 'Propulsor'
+        self.number_of_engines = 1.0
+        self.nacelle_diameter  = 1.0
+        self.engine_length     = 1.0
+        
+        #areas needed for drag; not in there yet
```
Comment (planes) [inline trunk/SUAVE/Components/Propulsors/Propulsor.py]: don't need the comment

File: trunk/SUAVE/Components/Energy/Converters/Generator_Zero_Fid.py (hunk 13 above)
Comment (WallyMaier) [reply trunk/SUAVE/Components/Energy/Converters/Generator_Zero_Fid.py]: ^^

File: trunk/SUAVE/Components/Energy/Converters/Generator_Zero_Fid.py (hunk 12 above)
Comment (WallyMaier) [reply trunk/SUAVE/Components/Energy/Converters/Generator_Zero_Fid.py]: ^^  try to remove the hardcoded values.....*Units.Seconds/Units.hour for 3600 i believe (double check)

**Pull request merged**